plutus-metatheory/src/VerifiedCompilation/UCaseOfCase.lagda.md

```diff
@@ -0,0 +1,100 @@
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip tt tt') -- recursive translation for the true branch
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip ft ft') -- recursive translation for the false branch
+                ----------------------
+                → X ⊢̂
+                       (case ((((force (builtin ifThenElse)) · b) · (constr tn tt)) · (constr fn ft)) alts)
+                       ⊳̂ ((((force (builtin ifThenElse)) · b) · (case (constr tn' tt') alts')) · (case (constr tn' tt') alts'))
+   var : ∀ {x : X} → X ⊢̂ (` x) ⊳̂ (` x) 
+   ƛ   : ∀ {x x' : Maybe X ⊢}
+           → Maybe X ⊢̂ x ⊳̂ x'
+           ----------------------
+           →  X ⊢̂ ƛ x ⊳̂ ƛ x' 
+   _·_ : ∀ {f t f' t' : X ⊢} → (X ⊢̂ f ⊳̂ f') → (X ⊢̂ t ⊳̂ t') → (X ⊢̂ (f · t) ⊳̂ (f' · t'))
+   force : ∀ {t t' : X ⊢} → X ⊢̂ t ⊳̂ t' → X ⊢̂ force t ⊳̂ force t'  
+   delay : ∀ {t t' : X ⊢} → X ⊢̂ t ⊳̂ t' → X ⊢̂ delay t ⊳̂ delay t'  
+   con : ∀ {tc : TmCon} → X ⊢̂ con tc ⊳̂ con tc
+   constr : ∀ {xs xs' : List (X ⊢)} { n : ℕ }
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip xs xs')
+                ------------------------
+                → X ⊢̂ constr n xs ⊳̂ constr n xs' 
+   case :  ∀ {p p' : X ⊢} {alts alts' : List (X ⊢)}
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip alts alts') -- recursive translation for the other case patterns
+                → X ⊢̂ p ⊳̂ p'
+                ------------------------
+                → X ⊢̂ case p alts ⊳̂ case p alts' 
+   builtin : ∀ {b : Builtin} → X ⊢̂ builtin b ⊳̂ builtin b
+   error : X ⊢̂ error ⊳̂ error
+
+```
+
+## Decision Procedure
+
+Since this compiler phase is just a syntax re-arrangement in a very particular situation, we can
+match on that situation in the before and after ASTs and apply the translation rule for this, or
+expect anything else to be unaltered.
+
+This can just traverse the ASTs applying the constructors from the transition relation.
+
+```
+--_⊢̂_⊳̂?_ : {X : Set} (ast ast' : X ⊢) → Dec (X ⊢̂ ast ⊳̂ ast')
+--_⊢̂_⊳̂?_ ast ast' = ?
+
+```
+
+## Semantic Equivalence
+
+We can show that this translation never alters the semantics of the statement. This is shown
+in terms of the CEK machine evaluation. Since it is a simple re-arrangement of the syntax, it
+isn't a refinement argument - the state before and after the operation is the same type, and is
+unaltered buy the syntax re-arrangement.
+
+This does rely on the encoding of the semantics of `IfThenElse` in the CEK module, since we
+need to show that the effective list of cases is the same as it would have been without the re-arrangement.
+
+The `stepper` function uses the CEK machine to evaluate a term. Here we call it with a very
+large gas budget and begin in an empty context (which assumes the term is closed).
+
+```
+--semantic_equivalence : ∀ {X set} {ast ast' : ⊥ ⊢}
```

Comment:
Bottom rather than top?

Reply:
They ought to be closed, so they should work with nothing in scope outside of themselves. However, maybe they need the three contract parameters?

Reply:
Right, silly me, it should be bottom of course. Sorry!

> However, maybe they need the three contract parameters?

I think not, but not really sure. IIRC some scripts take a different number of parameters, but don't quote me on that.

Reply:
James made the good point that the before and after things should reduce in the same way when treated as close terms. If they subsequently get applied to stuff I don't think that matters to them being equivalent as-is?

Reply:
> James made the good point that the before and after things should reduce in the same way when treated as close terms. If they subsequently get applied to stuff I don't think that matters to them being equivalent as-is?

I'd expect for some reasonable operational semantics to exist for functions that aren't applied. Some kind of bisimulation relation or behavioral equivalence or whatever. 

"For any given arguments these two functions return the same result" for example.

Reply:
This is what a sensibly abstract reduction semantics would do. Apparently there was one but it was ... "lost"... I shall look behind the sofa, and then might end up making a new one.

Reply:
> I'm pretty sure that I heard some mention of a proposal to make sure that all scripts take the same number of parameters.

See https://github.com/cardano-foundation/CIPs/pull/784#issuecomment-2078414218.